### PR TITLE
rpc: Fix wallet unload during walletpassphrase timeout

### LINKED ISSUE #14452: Unloading a wallet with in timeframe that is unlocked for crashes RPC server without any log/trace
Describe the issue 

For example, you unlock a wallet with "walletpassphrase" command for next 10 seconds, but for some reason unload wallet command is called before the wallet is locked again, it causes RPC server to crash and I didn't find any relevant help from debug.log file

This is important because whole idea of dynamically loading/unloading wallet is to help build applications that can work with multiple wallets more effectively, while its easy for an application to load all and every wallet but in our case we want to unload wallet right after necessary commands have passed (i.e. in PHP OOP I have registered a shutdown function for all wallets instances to automatically unload)

 What behavior did you expect?

unload command to return an error message in such circumstances

How reliably can you reproduce the issue, what are the steps to do so? 

1. Call "walletpassphrase" to unlock wallet for next N seconds
2. Issue "unloadwallet" command before those N seconds have passed

### PR DESCRIPTION
**Pull request by promag**
Replaces the raw wallet pointer in the `RPCRunLater` callback with a `std::weak_ptr` to check if the wallet is not expired.

To test:
```
bitcoind -regtest
bitcoin-cli -regtest encryptwallet foobar
bitcoin-cli -regtest walletpassphrase foobar 5 && bitcoin-cli -regtest unloadwallet ""
```

Fixes #14452.

### CODE REVIEW

**Comment by promag** [comment]
@furqansiddiqui please confirm if this change fixes your issue.

**Review by DocOBITCOIN — APPROVED** [review]

File: src/wallet/rpcwallet.cpp [hunk 1]
```diff
@@ -2004,7 +1997,14 @@ static UniValue walletpassphrase(const JSONRPCRequest& request)
     pwallet->TopUpKeyPool();
 
     pwallet->nRelockTime = GetTime() + nSleepTime;
-    RPCRunLater(strprintf("lockwallet(%s)", pwallet->GetName()), std::bind(LockWallet, pwallet), nSleepTime);
+    std::weak_ptr<CWallet> weak_wallet = wallet;
+    RPCRunLater(strprintf("lockwallet(%s)", pwallet->GetName()), [weak_wallet] {
+        if (auto wallet = weak_wallet.lock()) {
```
Comment (practicalswift) [inline src/wallet/rpcwallet.cpp]: Use another name here to avoid shadowing existing local variable `wallet` :-)

Comment (promag) [reply src/wallet/rpcwallet.cpp]: `wallet` is not captured.

Comment (practicalswift) [reply src/wallet/rpcwallet.cpp]: Sorry, sloppy wording from me - it is not a local variable. Still better to minimize surprises by not re-using the variable name, no?

Comment (promag) [reply src/wallet/rpcwallet.cpp]: What surprise?

**Comment by sipa** [comment]
utACK ada40e9a6cfb3d5d738658c739e80b03a86a2384

**Review by ryanofsky — APPROVED** [review]
utACK ada40e9a6cfb3d5d738658c739e80b03a86a2384

**Comment by ryanofsky** [comment]
From the description in #14452, it seems like it wouldn't be difficult to add a test (maybe in wallet_encryption.py) that triggers the current crash. But I think this PR looks good even without the test.

**Comment by promag** [comment]
Renamed variable and added a comment with rationale.

**Comment by ryanofsky** [comment]
utACK e24902bd4b2ad79e929cc36a67bc7d1edcec5d16

File: src/wallet/rpcwallet.cpp (hunk 1 above)
Comment (promag) [reply src/wallet/rpcwallet.cpp]: @practicalswift renamed.

**Comment by promag** [comment]
@ryanofsky ops e24902b has a typo. Fixed and squashed.

**Review by ryanofsky — COMMENTED** [review]
utACK 321decffa1fbf213462d97e5372bd0c4eeb99635. Only changes are shadow variable rename and new comment.

**Comment by promag** [comment]
Will add a test.

**Comment by promag** [comment]
@ryanofsky @jonasschnelli the test change fails in master.

**Comment by sipa** [comment]
reutACK 07a3942c51e0b55952bebe03f4c70a2d8f967c9c

**Comment by MarcoFalke** [comment]
utACK 07a3942c51e0b55952bebe03f4c70a2d8f967c9c

**Comment by promag** [comment]
@sipa @MarcoFalke sorry, just added a comment to the test.

**Comment by MarcoFalke** [comment]
re-utACK 8907df9e02ec47ef249a7422faa766f06aa01e94. Only change is a new comment in the test commit

**Review by ryanofsky — APPROVED** [review]
utACK 8907df9e02ec47ef249a7422faa766f06aa01e94. Only change is new test.

File: test/functional/wallet_multiwallet.py [hunk 2]
```diff
@@ -267,7 +268,11 @@ def wallet_file(name):
         assert 'w1' not in self.nodes[0].listwallets()
 
         # Successfully unload the wallet referenced by the request endpoint
+        # Also ensure unload works during walletpassphrase timeout
+        w2.encryptwallet('test')
+        w2.walletpassphrase('test', 1)
         w2.unloadwallet()
+        time.sleep(1.1)
```
[the remark this answers is not in the document]
Comment (promag) [reply test/functional/wallet_multiwallet.py]: Yeah, the sleep is not actually needed. However if the test terminates (and the node quits) then it's not guaranteed the timeout callback was executed? I can add a comment like "wait to ensure walletpassphrase callback is executed".

**Pull request merged**